File: maui/KsDialogs.Maui/Presentation/Toast.cs
using System;
using Microsoft.Maui.Controls;

namespace KsDialogs;

/// <summary>
/// Toast 表示の既定エントリ。
/// </summary>
/// <remarks>
/// <see cref="Instance"/> で手軽に呼び出せるほか、<c>new Toast()</c> を <see cref="IKsToast"/> として
/// DI 注入しても同じレジストリ (<see cref="ToastViewRegistry.Shared"/>) と同じ設定を共有する
/// (core/ADR-0002)。
/// <para>
/// この型は表示中のリストも計時も持たず、すべての呼び出しを委譲面越しに Native ライブラリの
/// coordinator へ渡す。そのため既定 singleton と注入したインスタンスを混ぜて使っても、
/// 重なり順と消滅の管理は 1 か所にまとまる。
/// </para>
/// </remarks>
public sealed class Toast : IKsToast
{
    // 既定エントリの生成は platform の委譲面の用意を伴うため、最初に使われるまで遅らせる
    private static readonly Lazy<Toast> s_instance = new(() => new Toast());

    private readonly ToastSettings _settings;
    private readonly IToastGateway _gateway;

    /// <summary>既定のレジストリと platform の委譲面を使うインスタンスを作る。</summary>
    public Toast() : this(ToastViewRegistry.Shared, ToastSettings.Shared, ToastGatewayFactory.Create())
    {
    }

    internal Toast(IToastGateway gateway)
        : this(ToastViewRegistry.Shared, ToastSettings.Shared, gateway)
    {
    }

    internal Toast(ToastViewRegistry registry, ToastSettings settings, IToastGateway gateway)
    {
        Registry = registry;
        _settings = settings;
        _gateway = gateway;
    }

    /// <summary>既定の singleton エントリ。</summary>
    public static Toast Instance => s_instance.Value;

    /// <inheritdoc/>
    public ToastViewRegistry Registry { get; }

    /// <inheritdoc/>
    public ToastStyle Style
    {
        get => _settings.Style;
        set
        {
            ArgumentNullException.ThrowIfNull(value);

            _settings.SetStyle(value, _gateway);
        }
    }

    /// <inheritdoc/>
    public void Show(string message, int? durationMs = null, DialogPlacement? placement = null)
    {
        ArgumentNullException.ThrowIfNull(message);

        _gateway.Show(ToastPresenter.Builtin(message, durationMs, placement));
    }

    /// <inheritdoc/>
    public void Show(IToastViewModel viewModel, int? durationMs = null, DialogPlacement? placement = null)
    {
        ArgumentNullException.ThrowIfNull(viewModel);

        _gateway.Show(ToastPresenter.Resolve(viewModel, Registry, durationMs, placement));
    }

    /// <inheritdoc/>
    public void Show<TViewModel>(
        TViewModel viewModel,
        Func<TViewModel, View> factory,
        int? durationMs = null,
        DialogPlacement? placement = null)
        where TViewModel : class, IToastViewModel
    {
        ArgumentNullException.ThrowIfNull(viewModel);
        ArgumentNullException.ThrowIfNull(factory);

        _gateway.Show(ToastPresenter.ResolveInline(
            viewModel,
            ToastViewFactories.Erase(factory),
            durationMs,
            placement));
    }

    /// <inheritdoc/>
    public void Show<TViewModel>(
        Action<TViewModel>? configure = null,
        int? durationMs = null,
        DialogPlacement? placement = null)
        where TViewModel : class, IToastViewModel =>
        // 解決だけを呼び出し時点で行い、生成と configure は中身の生成と同じ受理後の UI スレッドへ送る
        _gateway.Show(ToastPresenter.ResolveTyped(Registry, configure, durationMs, placement));
}

File: android/ksdialogs/src/main/kotlin/jp/kamusoft/ksdialogs/Toast.kt
package jp.kamusoft.ksdialogs

import android.content.Context
import android.view.View
import kotlin.reflect.KClass

/**
 * Toast 表示の既定エントリ。
 *
 * [Toast.instance] で手軽に呼び出せるほか、`Toast()` を [KsToast] として DI 注入しても
 * 同じ状態 (レジストリ・一括設定・表示中のリスト) を共有する (core/ADR-0002)。
 *
 * この型は状態を持たず、すべての呼び出しをプロセス内で唯一の coordinator へ委譲する。
 */
public class Toast internal constructor(
    internal val coordinator: ToastCoordinator,
) : KsToast {

    /** 既定の状態・レジストリ・取り付け先解決を使うインスタンスを作る。 */
    public constructor() : this(ToastCoordinator.shared)

    override val registry: ToastViewRegistry
        get() = coordinator.registry

    override var style: ToastStyle
        get() = coordinator.settings.style
        set(value) {
            coordinator.settings.style = value
        }

    override fun show(message: String, durationMs: Int?, placement: DialogPlacement?) {
        coordinator.accept(ToastContentRequest.Builtin(message), durationMs, placement)
    }

    override fun show(viewModel: ToastViewModel, durationMs: Int?, placement: DialogPlacement?) {
        coordinator.accept(ToastContentRequest.Registered(viewModel), durationMs, placement)
    }

    override fun <VM : ToastViewModel> show(
        viewModel: VM,
        durationMs: Int?,
        placement: DialogPlacement?,
        factory: Context.(VM) -> View,
    ) {
        // 型消去した factory をそのまま状態の正へ渡す。レジストリは経由しない (core/ADR-0013)
        coordinator.accept(
            ToastContentRequest.Inline(viewModel, erasedToastViewFactory(factory)),
            durationMs,
            placement,
        )
    }

    override fun <VM : ToastViewModel> show(
        viewModelClass: KClass<VM>,
        durationMs: Int?,
        placement: DialogPlacement?,
        configure: ((VM) -> Unit)?,
    ) {
        viewModelClass.requireReferenceTypeViewModel()
        // 解決は呼び出し時点のエントリのスナップショットで行い、以後の再登録には影響されない。
        // 未登録は受理そのものの失敗として、この場で同期に返す (core/ADR-0035)
        val entry = coordinator.registry.entry(viewModelClass)
        val viewModelFactory = entry?.viewModelFactory
            ?: throw DialogException.ViewModelFactoryNotRegistered(viewModelClass.viewModelTypeName)
        val viewFactory = entry.viewFactory
            ?: throw DialogException.ViewFactoryNotRegistered(viewModelClass.viewModelTypeName)

        coordinator.accept(
            ToastContentRequest.Typed(
                prepare = {
                    // ViewModel factory はクラス参照と対で登録されるため、生成物の型は常に一致する
                    @Suppress("UNCHECKED_CAST")
                    val created = viewModelFactory.createViewModel() as VM
                    configure?.invoke(created)
                    created
                },
                factory = viewFactory,
            ),
            durationMs,
            placement,
        )
    }

    public companion object {
        /** 既定の singleton エントリ。 */
        public val instance: Toast = Toast()
    }
}
